starknet-replay/src/storage/rpc/rpc_client.rs

```diff
@@ -0,0 +1,429 @@
+//! This module uses the Starknet RPC protocol to query the data from the
+//! Starknet RPC server.
+
+use starknet_api::block::{
+    BlockHash,
+    BlockHeader,
+    BlockTimestamp,
+    GasPrice,
+    GasPricePerToken,
+    StarknetVersion,
+};
+use starknet_api::core::{
+    ChainId,
+    ClassHash,
+    ContractAddress,
+    GlobalRoot,
+    Nonce,
+    SequencerContractAddress,
+};
+use starknet_api::data_availability::L1DataAvailabilityMode;
+use starknet_api::hash::StarkHash;
+use starknet_api::state::StorageKey;
+use starknet_api::transaction::{Transaction, TransactionReceipt};
+use starknet_core::types::{
+    BlockId,
+    ContractClass,
+    Felt,
+    MaybePendingBlockWithReceipts,
+    MaybePendingBlockWithTxHashes,
+    StarknetError,
+};
+use starknet_providers::jsonrpc::HttpTransport;
+use starknet_providers::{JsonRpcClient, Provider, ProviderError};
+use tokio::sync::OnceCell;
+use tracing::trace;
+use url::Url;
+
+use crate::block_number::BlockNumber;
+use crate::contract_address::to_field_element;
+use crate::error::DatabaseError;
+use crate::runner::replay_class_hash::ReplayClassHash;
+use crate::storage::rpc::receipt::convert_receipt;
+use crate::storage::rpc::transaction::convert_transaction;
+
+/// This structure partially implements a Starknet RPC client.
+///
+/// The RPC calls included are those needed to replay transactions.
+/// Clone is not derived because it's not supported by Client.
+pub struct RpcClient {
+    /// The endpoint of the Starknet RPC Node.
+    endpoint: Url,
+
+    /// The chain id variable initialised with the first call to
+    /// [`RpcClient::starknet_get_chain_id`]. This is doable because it's not
+    /// possible to replay blocks from different chains.
+    chain_id: OnceCell<ChainId>,
+
+    /// The most recent block number available from the rpc server. It is
+    /// initialised in [`RpcClient::starknet_block_number`]. The assumption
+    /// is that the latest blocks generated after the replay is started are
+    /// ignored.
+    block_number: OnceCell<BlockNumber>,
+}
+impl RpcClient {
+    /// Constructs a new `RpcStorage`.
+    ///
+    /// # Arguments
+    ///
+    /// - `endpoint`: The Url of the Starknet RPC node.
+    #[must_use]
+    pub fn new(endpoint: Url) -> Self {
+        RpcClient {
+            endpoint,
+            chain_id: OnceCell::new(),
+            block_number: OnceCell::new(),
+        }
+    }
+
+    /// This function generates a new client to perform an RPC request to the
+    /// endpoint.
+    ///
+    /// The client can't be shared across threads.
+    fn get_new_client(&self) -> JsonRpcClient<HttpTransport> {
+        JsonRpcClient::new(HttpTransport::new(self.endpoint.clone()))
+    }
+
+    /// This function queries the number of the most recent Starknet block.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`Err`] if the request fails.
```

Comment:
Does it return a specific variant of `DatabaseError`? Or multiple variants under different conditions?